src/wp-includes/user.php

```diff
@@ -425,10 +437,17 @@ function wp_authenticate_application_password( $input_user, $username, $password
 	$hashed_passwords = WP_Application_Passwords::get_user_application_passwords( $user->ID );
 
 	foreach ( $hashed_passwords as $key => $item ) {
-		if ( ! wp_check_password( $password, $item['password'], $user->ID ) ) {
+		$valid = wp_check_password( $password, $item['password'], $user->ID );
+
+		if ( ! $valid ) {
 			continue;
 		}
 
+		if ( wp_password_needs_rehash( $item['password'] ) ) {
```

Comment:
Independently off: 

https://github.com/johnbillion/wordpress-develop/pull/7

I don't think this is needed.

Application passwords are high-entropy.

Rehashing them from phpass to bcrypt is not needed, and not desired as its very slow on the validation part.

Reply:
I see your point. The idea was to opportunistically upgrade from phpass to bcrypt (blake2d in https://github.com/johnbillion/wordpress-develop/pull/7) to take advantage of the added preimage resistance. Given the known high entropy, perhaps that's not needed, but it's worth bearing in mind that application passwords are valid permanently until revoked.

Reply:
A fast hash is enough for application passwords.

They have 24 bit entropy, IIRC. The relevant threat model is SQL injection, so you don't want plaintext in the DB.

But fast hashes are fine.

I'd not add any upgrading of app passwords tbh, the less moving parts the better.

Reply:
All good points, thanks. I'll give it some more thought and likely remove the opportunistic hash upgrade for application passwords in https://github.com/johnbillion/wordpress-develop/pull/7.

Reply:
I left some comments on that PR. :)